### ISSUE TITLE
Outdated lockfile hides test failures

### ISSUE THREAD

**Issue by sneakers-the-rat** [linkml/linkml#1641]
**Describe the bug**
The CI installs the package using poetry, which respects the poetry lockfile.
not all installation methods respect the lockfile, and will install the most recent versions of packages that are allowed by the version pinning. This causes the tests to be overconfident. 

eg. Pydantic:

- lockfile pins 1.10.8: https://github.com/linkml/linkml/blob/0d2f3f0d5d65e619a00c81de81a5d1879e3c6830/poetry.lock#L1672
- dependency spec allows '*':  https://github.com/linkml/linkml/blob/0d2f3f0d5d65e619a00c81de81a5d1879e3c6830/pyproject.toml#L115

so a relock currently fetches 2.3.0 which causes the tests to fail because it doesn't allow keyword initialization in class MROs as used by `ConfiguredBaseModel` in the pydanticgen and tests. 

**To reproduce**

```
poetry lock
poetry install
pytest tests/test_generators/test_pydanticgen.py
```

**Expected behavior**
Should tighten up dependency specifiers or add versions to the test matrix

having a slow lockfile is good for keeping production systems stable but i don't think it's as good for in-use code in a language environment where lockfiles are not universally supported, unfortunately :(

**Comment by kevinschaper** [comment]
I wonder if we can use `pydantic = <2.0.0` as a dev dependency while keeping `pydantic = "*"` as a runtime dependency. I think that accurately describes our situation, but I don't know if it's supported.

**Comment by sneakers-the-rat** [comment]
I am one of those "upper bounds are good" ppl, which i know is controversial, but i think pydantic v2 is a decent example of when that is good - it isn't strictly true that the package is compatible with pydantic version `'*'`, since we don't know what v3 will be like!

So my opinions aside, if y'all want to keep the wide open dependency spec, I think the best case would be to add some explicit versions to the test matrix to run the tests for v1 and v2. 

I think tox can handle this ( eg. https://stackoverflow.com/q/59377071/13113166 ), or else you could do it with the github action w/ environment variables+poetry dep specifications or manual pip overrides to install specific versions. I could make a PR for that fi y'all are interested.

Other things that might be good to avoid getting surprised by this problem in the future
- More frequently re-lock, either before every commit or as part of every PR
- Add an additional testing branch to test both installation via pip (which the majority of ppl will probably do in practice) as well as poetry

Also I'm planning on submitting a PR to fix the pydantic problems here specifically, but i figured this was a more general problem than just the pydantic stuff (eg. i was in the process of handling https://github.com/linkml/linkml/issues/1567 , writing a test that failed before patching the bug to avoid future regressions, when i noticed that the existing tests should have been failing but werent because of the pinned version of rdflib. relocking causes those tests to correctly fail)

**Comment by cmungall** [comment]
@sneakers-the-rat - agreed. While we can't test the combinatorial explosion of every library version, pydantic is a special case. We should do both 1 and 2 in the matrix.

I also think more frequently re-locking is good. I am not so sure about in every PR, but definitely more frequent

**Comment by sneakers-the-rat** [comment]
up to y'all of course! poetry lockfiles sort of have an odd status at the moment because other package managers don't support them (and thus install the most recent matching packages compatible with the pyproject.toml file), so like where with npm/yarn the lockfile is more of a stable artifact of the package, with poetry/python most people install through pip or some variant and so don't see the lockfile. I think a very low-effort sanity check would be to do the thing mentioned above of just running the tests with both pip install and poetry installs with no additional labor in maintenance, but with a bit of a heads up if things are breaking.

also i just saw this example of how pandera does multi-version testing fwiw: https://github.com/unionai-oss/pandera/blob/main/.github/workflows/ci-tests.yml

**Comment by sierra-moxon** [comment]
from our Dev call today:
- we likely won't support non-poetry python env's for development of LinkML packages
- folks can create a requirements.txt (that echoes the poetry.lock file versions) using this poetry command: `poetry export -f requirements.txt --output requirements.txt` (though if you have poetry installed already, but want to use a different env manager, this might seem a bit strange :))
- we can try a limited number of matrixed package testing strategies.

@sneakers-the-rat  - shall we close this ticket?  :)

**Comment by sneakers-the-rat** [comment]
fair enough, though tbc this specifically breaks non-development installs - it breaking `pip install linkml` and the development install `poetry install` not matching non-development conditions is the problem.

Avoiding people needing to generate a `requirements.txt` is a good goal imo, since that's an awful lot of overhead compared to `pip install linkml`, and also makes it pretty challenging to depend on it and develop downstream packages with something that isn't poetry.

I do think that it would be the lowest effort/most direct fix to just add `pydantic_version = ["<2.0.0", ">=2.0.0"]` to the matrix and add a `poetry run pip install pydantic${{ matrix.pydantic-version }}` step after `poetry install` step, at least until the dust clears/y'all drop support for 1 and ppl mostly move to 2. as the timheidecker put it, "its free test coverage," and accurately reflects the dependency spec (where currently specifying `'*'` and only testing against the version in the lockfile is technically inaccurate)

up 2 u to close, for now i've just monkeypatched almost the whole generator to fix some problems like this until i can put together a pull request so it no longer affects me.
